src/LayerEditor/ui/data/block_item.py

```diff
@@ -106,31 +119,31 @@ def init_regions_for_new_shape(self, shape_dim, shape_id):
                 else:
                     layer.set_region_to_shape(shape_dim, shape_id, RegionItem.none)
 
-    #TODO: make this undoable
-    def insert_layer(self, layer_data, index):
-        pass
-        # layer = LayerModel(self.decomposition, self.regions_model.regions)
-        # self.layers.insert(layer)
-        # self.layers_dict.add(layer)
-        #
-        #
-        #
-        # if len(self.layers) == 1:
-        #     for shape in [*self.decomposition.points,
-        #                   *self.decomposition.segments,
-        #                   *self.decomposition.polygons]:
-        #         shape.attr[layer] = shape.attr[self.layers[-2]]
-        # else:
-        #     for shape in [*self.decomposition.points,
-        #                   *self.decomposition.segments,
-        #                   *self.decomposition.polygons]:
-        #         shape.attr[layer] = 0
-        #
-        # self.gui_selected_layer = self.layers[0]
-
-    def save(self):
-        """Save data from this block to LayerGeometryModel"""
-        layers = []
-        for layer in self.layers:
-            layers.append(layer.save())
-        return layers
+    @undo.undoable
+    def add_layer(self, new_layer: LayerItem):
+        old_block = new_layer.block
+        self.layers_dict.add(new_layer)
+        new_layer.block = self
+        yield "Add Layer"
+        self.delete_layer(new_layer)
+        new_layer.block = old_block
```

Comment:
That is the intended way to do things. and its true that it is not necesary to preserve `old_block` because it should be always None. It was replaced by assert which checks if `new_layer.block is None`.